src/edu/sjsu/stratagem/Environment.java

```diff
@@ -3,9 +3,13 @@
 import java.util.Map;
 import java.util.HashMap;
 
-public class Environment {
-    private Map<String,Value> env = new HashMap<String,Value>();
-    private Environment outerEnv;
+/**
+ * A variable environment for the program. During typechecking of a program V
+ * will be Type, and during evaluation of a program V will be Value.
+ */
+public class Environment<V> {
```

Comment:
Unfortunately, I decided to split this up into two hardcoded classes while implementing the two StratagemException classes above. Java's generics weren't powerful enough.

Depending on whether an `Environment` is specialized with `Value` or `Type`, we need to choose between throwing a `StratagemRuntimeException` or a `StratagemTypecheckException`. We throw these exceptions when looking up a non-existent variable or when creating a duplicate variable. In C++ (a language I am more familiar with), I would use partial template specialization and write separate variable lookup functions for `Environment<Type>` and `Environment<Value>`, but in Java I changed `Environment<V>` to `Environment<V extends EnvironmentType>`, where `EnvironmentType` looks like the following:

```java
interface EnvironmentType {
    EnvironmentType makeNew();
    void throwException(String message) throws StratagemException;
}
```

which delegates the responsibility of throwing the correct exception type to our template parameter type. We then make the `Value` and `Type` classes implement `EnvironmentType`.

Within `EnvironmentType`, the `throwException` method is to throw the right exception type at runtime and the `makeNew` is merely technical baggage because we need an instance of an `EnvironmentType` (and instance of  a `Value` or `Type`) to call `throwException` since Java interfaces can't have static methods.

I felt this was more complex than was warranted (I was simply fighting the Java language at this point), so I decided to simply un-template-ize `Environment` into two classes. Additionally, if we had continued with this route it might have caused problems when we try to include more information in Stratagem error messages later on, such as location of the error, etc.

I'm kind of disappointed in Java. Seems like the type of thing Scala would be better at. Oh well, we'll live with the code duplication for now!